reactor-netty-http/src/main/java/reactor/netty/http/client/Http2Pool.java

```diff
@@ -0,0 +1,829 @@
+	}
+
+	@Override
+	public long secondsSinceLastInteraction() {
+		long sinceMs = clock.millis() - lastInteractionTimestamp;
+		return sinceMs / 1000;
+	}
+
+	@Override
+	public Mono<Integer> warmup() {
+		return Mono.just(0);
+	}
+
+	void cancelAcquire(Borrower borrower) {
+		if (!isDisposed()) {
+			ConcurrentLinkedDeque<Borrower> q = pending;
+			q.remove(borrower);
+		}
+	}
+
+	Mono<Void> destroyPoolable(Http2PooledRef ref) {
+		Mono<Void> userProvidedDestroy;
+		try {
+			userProvidedDestroy = Mono.from(destroyHandler().apply(this, ref));
+		}
+		catch (Throwable destroyFunctionError) {
+			userProvidedDestroy = Mono.error(destroyFunctionError);
+		}
+		return userProvidedDestroy;
+	}
+
+	BiFunction<Http2Pool, Http2PooledRef, Publisher<Void>> destroyHandler() {
+		return (pool, pooledRef) -> {
+			pooledRef.slot.decrementConcurrency();
+			Connection connection = pooledRef.poolable();
+			Http2FrameCodec frameCodec = connection.channel().pipeline().get(Http2FrameCodec.class);
+			if (frameCodec != null) {
+				if (pooledRef.slot.concurrency() == 0 && pooledRef.slot.invalidate()) {
+					releaseConnection(connection);
+				}
+			}
+			else {
+				// no need to guard here with invalidate()
+				// #findConnection will never give a connection that does not have Http2FrameCodec
+				poolConfig.allocationStrategy().returnPermits(1);
+				removeSlot(pooledRef.slot);
+			}
+			return Mono.empty();
+		};
+	}
+
+	void doAcquire(Borrower borrower) {
+		if (isDisposed()) {
+			borrower.fail(new PoolShutdownException());
+			return;
+		}
+
+		pendingOffer(borrower);
+		drain();
+	}
+
+	void drain() {
+		if (WIP.getAndIncrement(this) == 0) {
+			drainLoop();
+		}
+	}
+
+	void drainLoop() {
+		recordInteractionTimestamp();
+		int maxPending = poolConfig.maxPending();
+
+		for (;;) {
+			@SuppressWarnings("unchecked")
+			ConcurrentLinkedQueue<Slot> resources = CONNECTIONS.get(this);
+			@SuppressWarnings("unchecked")
+			ConcurrentLinkedDeque<Borrower> borrowers = PENDING.get(this);
+			if (resources == null || borrowers == TERMINATED) {
+				return;
+			}
+
+			int borrowersCount = borrowers.size();
+
+			if (borrowersCount != 0) {
+				// find a connection that can be used for opening a new stream
+				Slot slot = findConnection(resources);
+				if (slot != null) {
+					Borrower borrower = borrowers.pollFirst();
+					if (borrower == null) {
+						resources.offer(slot);
+						continue;
+					}
+					if (isDisposed()) {
+						borrower.fail(new PoolShutdownException());
+						return;
+					}
+					if (slot.activate()) {
+						borrower.stopPendingCountdown();
+						ACQUIRED.incrementAndGet(this);
+						slot.incrementConcurrency();
+						// we are ready here, the connection can be used for opening another stream
+						slot.deactivate();
+						poolConfig.acquisitionScheduler().schedule(() -> borrower.deliver(new Http2PooledRef(slot)));
+					}
+					else {
+						borrowers.offerFirst(borrower);
+						continue;
+					}
+				}
+				else {
```

Comment:
I need to look back at this part of the drain loop more in details. one thing I noticed is that it implements warmup.

Reply:
nevermind the warmup comment, this was explained in header javadoc